File: server/src/sync/machineCache.ts
import { z } from 'zod'
import type { Store } from '../store'
import { clampAliveTime } from './aliveTime'
import { EventPublisher } from './eventPublisher'

const machineMetadataSchema = z.object({
    host: z.string().optional(),
    platform: z.string().optional(),
    happyCliVersion: z.string().optional(),
    displayName: z.string().optional()
}).passthrough()

export interface Machine {
    id: string
    namespace: string
    seq: number
    createdAt: number
    updatedAt: number
    active: boolean
    activeAt: number
    metadata: {
        host: string
        platform: string
        happyCliVersion: string
        displayName?: string
        [key: string]: unknown
    } | null
    metadataVersion: number
    daemonState: unknown | null
    daemonStateVersion: number
}

export class MachineCache {
    private readonly machines: Map<string, Machine> = new Map()
    private readonly lastBroadcastAtByMachineId: Map<string, number> = new Map()

    constructor(
        private readonly store: Store,
        private readonly publisher: EventPublisher
    ) {
    }

    getMachines(): Machine[] {
        return Array.from(this.machines.values())
    }

    getMachinesByNamespace(namespace: string): Machine[] {
        return this.getMachines().filter((machine) => machine.namespace === namespace)
    }

    getMachine(machineId: string): Machine | undefined {
        return this.machines.get(machineId)
    }

    getMachineByNamespace(machineId: string, namespace: string): Machine | undefined {
        const machine = this.machines.get(machineId)
        if (!machine || machine.namespace !== namespace) {
            return undefined
        }
        return machine
    }

    getOnlineMachines(): Machine[] {
        return this.getMachines().filter((machine) => machine.active)
    }

    getOnlineMachinesByNamespace(namespace: string): Machine[] {
        return this.getMachinesByNamespace(namespace).filter((machine) => machine.active)
    }

    getOrCreateMachine(id: string, metadata: unknown, daemonState: unknown, namespace: string): Machine {
        const stored = this.store.getOrCreateMachine(id, metadata, daemonState, namespace)
        return this.refreshMachine(stored.id) ?? (() => { throw new Error('Failed to load machine') })()
    }

    refreshMachine(machineId: string): Machine | null {
        const stored = this.store.getMachine(machineId)
        if (!stored) {
            const existed = this.machines.delete(machineId)
            if (existed) {
                this.publisher.emit({ type: 'machine-updated', machineId, data: null })
            }
            return null
        }

        const existing = this.machines.get(machineId)

        const metadata = (() => {
            const parsed = machineMetadataSchema.safeParse(stored.metadata)
            if (!parsed.success) return null
            const data = parsed.data as Record<string, unknown>
            const host = typeof data.host === 'string' ? data.host : 'unknown'
            const platform = typeof data.platform === 'string' ? data.platform : 'unknown'
            const happyCliVersion = typeof data.happyCliVersion === 'string' ? data.happyCliVersion : 'unknown'
            const displayName = typeof data.displayName === 'string' ? data.displayName : undefined
            return { host, platform, happyCliVersion, displayName, ...data }
        })()

        const storedActiveAt = stored.activeAt ?? stored.createdAt
        const existingActiveAt = existing?.activeAt ?? 0
        const useStoredActivity = storedActiveAt > existingActiveAt

        const machine: Machine = {
            id: stored.id,
            namespace: stored.namespace,
            seq: stored.seq,
            createdAt: stored.createdAt,
            updatedAt: stored.updatedAt,
            active: useStoredActivity ? stored.active : (existing?.active ?? stored.active),
            activeAt: useStoredActivity ? storedActiveAt : (existingActiveAt || storedActiveAt),
            metadata,
            metadataVersion: stored.metadataVersion,
            daemonState: stored.daemonState,
            daemonStateVersion: stored.daemonStateVersion
        }

        this.machines.set(machineId, machine)
        this.publisher.emit({ type: 'machine-updated', machineId, data: machine })
        return machine
    }

    reloadAll(): void {
        const machines = this.store.getMachines()
        for (const machine of machines) {
            this.refreshMachine(machine.id)
        }
    }

    handleMachineAlive(payload: { machineId: string; time: number }): void {
        const t = clampAliveTime(payload.time)
        if (!t) return

        const machine = this.machines.get(payload.machineId) ?? this.refreshMachine(payload.machineId)
        if (!machine) return

        const wasActive = machine.active
        machine.active = true
        machine.activeAt = Math.max(machine.activeAt, t)

        const now = Date.now()
        const lastBroadcastAt = this.lastBroadcastAtByMachineId.get(machine.id) ?? 0
        const shouldBroadcast = (!wasActive && machine.active) || (now - lastBroadcastAt > 10_000)
        if (shouldBroadcast) {
            this.lastBroadcastAtByMachineId.set(machine.id, now)
            this.publisher.emit({ type: 'machine-updated', machineId: machine.id, data: { activeAt: machine.activeAt } })
        }
    }

    expireInactive(now: number = Date.now()): void {
        const machineTimeoutMs = 45_000

        for (const machine of this.machines.values()) {
            if (!machine.active) continue
            if (now - machine.activeAt <= machineTimeoutMs) continue
            machine.active = false
            this.publisher.emit({ type: 'machine-updated', machineId: machine.id, data: { active: false } })
        }
    }
}

File: hub_go/internal/sync/machine_cache.go
package sync

import (
	"sync"
	"time"

	"hub_go/internal/store"
)

type MachineCache struct {
	store     *store.Store
	publisher *EventPublisher
	mu        sync.RWMutex
	machines  map[string]*Machine
}

func NewMachineCache(store *store.Store, publisher *EventPublisher) *MachineCache {
	return &MachineCache{
		store:     store,
		publisher: publisher,
		machines:  map[string]*Machine{},
	}
}

func (c *MachineCache) GetMachine(id string, namespace string) *Machine {
	c.mu.RLock()
	machine := c.machines[id]
	c.mu.RUnlock()
	if machine != nil {
		return machine
	}
	machine, _ = c.store.GetMachine(namespace, id)
	if machine == nil {
		return nil
	}
	c.mu.Lock()
	c.machines[id] = machine
	c.mu.Unlock()
	return machine
}

func (c *MachineCache) GetMachinesByNamespace(namespace string) []Machine {
	if c.store == nil {
		return nil
	}
	machines := c.store.ListMachines(namespace)
	c.mu.Lock()
	for i := range machines {
		machine := machines[i]
		c.machines[machine.ID] = &machine
	}
	c.mu.Unlock()
	return machines
}

func (c *MachineCache) HandleMachineAlive(payload MachineAlivePayload) {
	if c == nil || c.store == nil {
		return
	}
	aliveAt := clampAliveTime(payload.Time)
	if aliveAt == 0 {
		aliveAt = time.Now().UnixMilli()
	}
	machine, _ := c.store.UpsertMachine(payload.Namespace, payload.MachineID, nil, nil)
	if machine == nil {
		return
	}
	machine.Active = true
	machine.ActiveAt = aliveAt
	machine.UpdatedAt = time.Now().UnixMilli()
	_ = c.store.UpdateMachine(payload.Namespace, machine)
	c.mu.Lock()
	c.machines[payload.MachineID] = machine
	c.mu.Unlock()

	if c.publisher != nil {
		c.publisher.Emit(SyncEvent{
			Type:      "machine-updated",
			Namespace: payload.Namespace,
			MachineID: payload.MachineID,
			Data: map[string]any{
				"id":                 machine.ID,
				"namespace":          machine.Namespace,
				"createdAt":          machine.CreatedAt,
				"updatedAt":          machine.UpdatedAt,
				"metadata":           machine.Metadata,
				"metadataVersion":    machine.MetadataVersion,
				"runnerState":        machine.RunnerState,
				"runnerStateVersion": machine.RunnerStateVersion,
				"active":             machine.Active,
				"activeAt":           machine.ActiveAt,
				"seq":                machine.Seq,
			},
		})
	}
}
